### PR TITLE
HEEDLS-244 Tutorial next button back end

### PR DESCRIPTION
**Pull request by benjimarshall**
## Changes
Amend `GetTutorialInformation` query and data model to get the next
tutorial ID, the next section ID, and the post-learning assessment
path.

## Testing
- Add test for course contents section test: test a course which has some sections where none of the section's
tutorials are in the CustomisationTutorial table (based on some checks I did this morning)
- Amend `GetTutorialInformation` tests, and add tests for edge cases like being the last tutorial
in a section, and check tutorials not in CustomisationTutorials are skipped.

### CODE REVIEW

File: DigitalLearningSolutions.Data.Tests/Services/CourseContentServiceTests.cs [hunk 1]
```diff
@@ -445,6 +445,7 @@ public void Get_course_content_should_only_return_sections_with_tutorials_in_cus
         [TestCase(1, 15853, 101, 173218)]
         [TestCase(254480, 24224, 101, null)]
         [TestCase(22044, 10059, 121, 100467)]
+        [TestCase(11, 6226, 101, 89490)]
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data.Tests/Services/CourseContentServiceTests.cs:448]: Great, thank you! Do keep adding any course you think might be useful to this test, the more the better! 🙂

File: DigitalLearningSolutions.Data.Tests/Services/TutorialContentServiceTests.cs [hunk 2]
```diff
@@ -46,10 +46,88 @@ public void Get_tutorial_information_should_return_tutorial()
...
+        [Test]
+        public void Get_tutorial_information_nextTutorial_should_skip_tutorials_not_in_customisation()
+        {
+            // Given
+            const int candidateId = 210934;
+            const int customisationId = 18366;
+            const int sectionId = 973;
+            const int tutorialId = 4257;
+
+            // The next tutorial ID in this section is 4258, but the next tutorial selected in CustomisationTutorials is 4263
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data.Tests/Services/TutorialContentServiceTests.cs:99]: Very useful comment, thanks! I think these tests where we're relying on the state of the db is a case where comments can be really useful 🙂

File: DigitalLearningSolutions.Data/Services/TutorialContentService.cs [hunk 3]
```diff
@@ -49,17 +49,39 @@ int tutorialId
                          Tutorials.Objectives,
                          Tutorials.VideoPath,
                          Tutorials.TutorialPath,
-                         Tutorials.SupportingMatsPath AS SupportingMaterialPath
+                         Tutorials.SupportingMatsPath AS SupportingMaterialPath,
+                         Sections.PLAssessPath AS PostLearningAssessmentPath,
+                         MIN(SubsequentTutorials.TutorialID) AS NextTutorialID,
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data/Services/TutorialContentService.cs:54]: Does this return the next tutorial according to the OrderByNumber? 🙂

File: DigitalLearningSolutions.Data/Services/TutorialContentService.cs [hunk 4]
```diff
@@ -49,17 +49,39 @@ int tutorialId
                          Tutorials.Objectives,
                          Tutorials.VideoPath,
                          Tutorials.TutorialPath,
-                         Tutorials.SupportingMatsPath AS SupportingMaterialPath
+                         Tutorials.SupportingMatsPath AS SupportingMaterialPath,
+                         Sections.PLAssessPath AS PostLearningAssessmentPath,
+                         MIN(SubsequentTutorials.TutorialID) AS NextTutorialID,
+                         MIN(SubsequentSections.SectionID) AS NextSectionID
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data/Services/TutorialContentService.cs:55]: Does this return the next section according to the SectionNumber? 🙂

File: DigitalLearningSolutions.Data/Services/TutorialContentService.cs [hunk 5]
```diff
@@ -49,17 +49,39 @@ int tutorialId
...
+                         ON SubsequentCustomisationTutorials.CustomisationID = Customisations.CustomisationID
+                            AND SubsequentCustomisationTutorials.Status = 1
+
+                         LEFT JOIN Tutorials AS SubsequentTutorials
+                         ON SubsequentTutorials.SectionID = Sections.SectionID
+                            AND Tutorials.OrderByNumber < SubsequentTutorials.OrderByNumber
+                            AND SubsequentCustomisationTutorials.TutorialID = SubsequentTutorials.TutorialID
+
+                         LEFT JOIN Tutorials AS SubsequentSectionsTutorials
+                         ON  SubsequentCustomisationTutorials.TutorialID = SubsequentSectionsTutorials.TutorialID
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data/Services/TutorialContentService.cs:79]: Why are we joining on Tutorials twice? 🙂

**Review by FridaTveit — COMMENTED** [review]
A few comments, otherwise looks good 🙂

File: DigitalLearningSolutions.Data.Tests/Services/CourseContentServiceTests.cs (hunk 1 above)
Comment (benjimarshall) [reply DigitalLearningSolutions.Data.Tests/Services/CourseContentServiceTests.cs:448]: Will do 🙂

File: DigitalLearningSolutions.Data.Tests/Services/TutorialContentServiceTests.cs (hunk 2 above)
Comment (benjimarshall) [reply DigitalLearningSolutions.Data.Tests/Services/TutorialContentServiceTests.cs:99]: Yeah absolutely, and it was a bit ridiculous when I tried to encode all of this in the function title! 🙂

File: DigitalLearningSolutions.Data/Services/TutorialContentService.cs (hunk 5 above)
Comment (benjimarshall) [reply DigitalLearningSolutions.Data/Services/TutorialContentService.cs:79]: The `SubsequentTutorials` is used to join on the other tutorials in this section (that are in `CustomisationTutorials`) but only if it has a greater `OrderBy`. This is done in a left join because that will give us a single row with `null` in this `NextTutorialId`, instead of giving us an empty table if we join first and then filter later. There might be more than one tutorial that matches though, so we aggregate them into a single row with the `MIN` function.

The second join on tutorials, `SubsequentSectionsTutorials` gives us all of the tutorials in the customisation (which we then use to find all the sections that are in the customisation, as opposed to the customisations in the application), which we then join `SubsequentSections` to give us rows for each section in the customisation.

At one point I did just have the simpler second join, but I think it would make finding the next tutorial in the section more difficult, because if the tutorial is the last in the section we want this field to be `null`, and not just `WHERE` on having a greater `OrderBy` and getting an empty table. There might be another way to do this but I don't know about it.

Does that make sense? I'm not sure how clear that all was 🙂

File: DigitalLearningSolutions.Data/Services/TutorialContentService.cs (hunk 4 above)
Comment (benjimarshall) [reply DigitalLearningSolutions.Data/Services/TutorialContentService.cs]: Yes it does, I added that as a comment in bc4cce2 🙂

File: DigitalLearningSolutions.Data/Services/TutorialContentService.cs (hunk 3 above)
Comment (benjimarshall) [reply DigitalLearningSolutions.Data/Services/TutorialContentService.cs]: Yes it does, I added that as a comment in bc4cce2 🙂

File: DigitalLearningSolutions.Data/Services/TutorialContentService.cs (hunk 5 above)
Comment (FridaTveit) [reply DigitalLearningSolutions.Data/Services/TutorialContentService.cs]: >There might be more than one tutorial that matches though, so we aggregate them into a single row with the MIN function

So `SubsequentTutorials` gives all the subsequent tutorial in this section right? So if there's more than one tutorial after this one, how to we know that the MIN one is the next one? Shouldn't it be MIN based on the `OrderBy` instead of the `TutorialID`? 🙂 

The two joins on Tutorials make sense then, thanks 🙂

Comment (benjimarshall) [reply DigitalLearningSolutions.Data/Services/TutorialContentService.cs]: Yep this is a problem, thanks for catching it! 🙂

Comment (benjimarshall) [reply DigitalLearningSolutions.Data/Services/TutorialContentService.cs]: Fixed and regression tests added in 79d8087 🙂

**Comment by benjimarshall** [comment]
Something else I just thought of is that there are a few sections with different tutorials sharing `OrderByNumber`s. I suspect there also are sections sharing numbers in a course. Currently this will arbitrarily pick one of them, but this means pressing next continuously from start to end of a course could lead to some sections being missed. 

Should we leave this (because tutorials sharing `OrderByNumber`s sounds strange, and only ~750/13000 tutorials do so), or try to account for this in some way (maybe ordering by tutorial ID as a tie break)? What do you think? 🙂

**Comment by FridaTveit** [comment]
Good question... I'll check with Kevin if this is to be expected 🙂

**Review by FridaTveit — APPROVED** [review]
Looks good 🙂 Other than the concern about order by and section numbers being shared... But feel free to merge this for now and deal with that later if necessary 🙂

**Comment by benjimarshall** [comment]
0a8424f adds support for shared `OrderByNumber`s and `SectionNumber`s. There's no tests for shared section numbers because there are no examples of this occurring in the database.

**Review by FridaTveit — APPROVED** [review]
Looks good 🙂
